lib/fileset/default.nix

```diff
@@ -0,0 +1,580 @@
+          else if ! pathExists path then
+            throw "lib.fileset.singleton: Path ${toString path} does not exist"
+          else create (singletonInternal path (pathType path));
+
+        /*
+        Coerce a value to a fileset.
+        If the value is already a fileset, return it unchanged.
+        If the value is a path, create a fileset from the path, see `singleton`.
+
+        This function is effectively not needed because all functions in `lib.fileset` that expect a file set will implicitly apply this function.
+
+        Type:
+          coerce :: Any -> FileSet
+        */
+        coerce = value:
+          if value._type or "" == "fileset" then
+            value
+          else
+            singleton value;
+
+        /*
+        Trace all files included in the file set in a pretty way. Note that this strictly evaluates the file set. Example output:
+
+            trace: /home/user/src/myProject
+            trace: - Makefile (regular)
+            trace: - src
+            trace:   - main.c (regular)
+
+
+        # TODO: Make the output configurable
+
+        Type:
+          trace :: FileSet -> FileSet
+        */
+        trace = fileset':
+          let
+            fileset = coerce fileset';
+            result = x: x;
+
+            traceTree = indent: printParents: name: tree: value:
+              if tree == "directory" then
+                builtins.seq printParents (builtins.trace "${indent}- ${name} (recursive directory)" value)
+              else if isAttrs tree then
+                let
+                  printSelf = builtins.seq printParents (builtins.trace "${indent}- ${name}" null);
+                in
+                  foldl' (value: name':
+                    traceTree (indent + "  ") printSelf name' tree.${name'} value
+                  ) value (attrNames tree)
+              else if tree == null then
+                value
+              else
+                builtins.seq printParents (builtins.trace "${indent}- ${name} (${tree})" value);
+
+          in
+            # TODO: Don't use toString
+            if fileset.tree == null then
+              builtins.trace "${toString base}" result
+            else if fileset.tree == "directory" then
+              builtins.trace "${toString base} (recursive directory)" result
+            else if isAttrs fileset.tree then
+              builtins.trace "${toString base}" (
+                foldl' (value: name:
+                  traceTree "" null name fileset.tree.${name} value
+                ) result (attrNames fileset.tree)
+              )
+            else
+              builtins.trace "${toString base} (${fileset.tree})" result;
+
+        /*
+        Create a file set containing all files tracked by a local git repository
+
+        TODO: Make this configurable, we might need it in the future
+
+        Type:
+          gitTracked :: Path -> FileSet
+        */
+        gitTracked = path:
```

Comment:
This module is about the fileset data type and its operations. This code isn't about that at all; it is rather a consumer of the fileset module, just like git itself is a consumer of the OS file system.
A separate `git` module would be more appropriate, or you could add it later. It is somewhat useful for validating the fileset concept and implementation though.

Reply:
Sounds good, I wasn't sure if it could be implemented without having access to internals, but I figured out it's actually pretty easy using a file filter (currently called `matching`) with a predicate that checks whether the file is in the imported store path.

Reply:
There's a clear separation between core functions and extra convenience functions. The current state of this PR is that _only_ core functions are in `lib.fileset`. However, I do think `lib.fileset` would also be the right place for convenience functions.

In any case `gitTracked` is removed entirely for now. It can't be implemented efficiently without something like https://github.com/NixOS/nix/pull/6530 or https://github.com/NixOS/nix/issues/2944